### ISSUE TITLE
security: certificate verification only uses short hostname

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#4434]
While #4379 is being handled I tried to set up my certificates for AWS and run into the following:

```
W0217 15:18:23.772823 14548 rpc/client.go:326  x509: certificate is valid for localhost, ..., ip-172-31-52-143.ec2.internal, ..., not ip-172-31-52-143
I0217 15:18:24.599365 14548 server.go:1934  http: TLS handshake error from 172.31.61.86:46360: remote error: bad certificate
I0217 15:18:25.389928 14548 server.go:1934  http: TLS handshake error from 172.31.63.149:10953: EOF
```

Meanwhile the full hostname for that node is `ip-172-31-52-143.ec2.internal` and this is what I specified after `create-cert`. This looks like the cert verif code only uses the short hostname. This is invalid; the cert verification code should match either the IP address the node is bound to; or the full hostname.

**Comment by mberhault** [comment]
we don't control which host gets verified, so you need to specify all the names that can be used to reach the host when creating the cert.

**Comment by knz** [comment]
Let me clarify.
The name used to reach the host is properly present in the list carried by the certificate. This is what the message said. However this is not the issue.

The issue is that the name used as reference is whatever name was given to --host, even when this name is not fully qualified. This is a problem. It's both a practical problem and potential security issue. Practical: if the cert carries the fqdn then it's silly to refuse the equivalent hostname given to --host

Security: I can lie via --host.

What the code should do is either validate against the IP address it is listening to, or if that fails forward then reverse lookup the value of --host to get the real fqdn and validate against that. 
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by tamird** [comment]
This is essentially a dupe of
https://github.com/cockroachdb/cockroach/issues/1008

**Comment by knz** [comment]
@tamird not sure this is #1008 exactly. Here the problem is FQDN vs. non-qualified names. The cert check code should qualify the name used for the check as a matter of course, regardless of whether the node can be reached using that address or not.

**Comment by tamird** [comment]
I'm not following you here. What do you mean by "qualify"? qualify with what - all we know about the node is its hostname; what additional information would we use to augment the hostname?

**Comment by knz** [comment]
So basically "qualified" is a term used with host names. So if you have a host "foo" in a sub-network "bar"  of a search domain "microsoft.com", "foo" is unqualified, "foo.bar" is partially qualified and "foo.bar.microsoft.com" is fully qualified.

Now you can look at the error message I initially reported. The problem is that:

- the certificate properly contains the FQDN "ip-172-31-52-143.ec2.internal", as it should
- when I start the server without arguments, it picks up the unqualified name "ip-172-31-52-143" and tries to use this to check the certificate.

My complaint is that the cert check code should qualify the host name first.

**Comment by tamird** [comment]
I see. Seems that we'll want to use https://godoc.org/github.com/miekg/dns or similar for this.

**Comment by bdarnell** [comment]
We don't need a third-party dns package; `net.LookupAddr` should suffice (if reverse-dns lookups are what we need here). But I'm not sure if that's the right solution. 

Where did the two different hostnames here come from? It looks like the nodes can address each other using their unqualified hostname, so why did you create the certificate for the qualified version? If `os.Hostname` doesn't return the right thing, then I think we're going to need the admin to pass in the correct hostname with `--host` (or the new flag proposed in #1008); it's unlikely that we could derive the right name from DNS or other sources.

**Comment by knz** [comment]
@bdarnell thanks for asking :-)
So here we're looking at the AWS scenario. When you create an aws  cluster using either tool we advertise, the tool informs you only about the public IP address of your newly created nodes. If like me you're an innocent user you try to set up your cockroach nodes using this information and you soon hit one of the known issues. So reading these issues' discussions you are instructed to use the local address not the public address. But you don't have this information yet. 

There are two ways to do this. One is the "hostname" command. But since you're likely to see and use the local IP addresses too, you're likely to use ifconfig instead of "hostname" then reverse lookup the local IP address to get the local name in FQDN form. 

Because that's what should work, I mean any sysadm learned this should yield equivalent results.

Moreover security best practices always suggest using the FQDN because it reduces ambiguity.

So there you are with an valid, AWS-provided FQDN for your local address, and you go on to test that: you start 'cockroach start --insecure' without arguments, seems to work. You then use the newly known local FQDN from another nodes, and that seems to work too!

Then you try again without --insecure and it doesn't work anymore.

And then you start struggling because it looks like you've done everything right to this point, the docs don't say you should have done something else, and yet it doesn't work.

Tldr: a local address will resolve to a FQDN  when the user asks DNS and the user will use that. Meanwhile many standard setups only populate the unqualified hostname in is.Hostname. we should deal with it. For the sake of usability, and probably security as well.

**Comment by bdarnell** [comment]
Can we fix the tools we provide to give you the internal IP/hostname instead of the public one?

Concretely here's what's happening on an ubuntu 14.04 instance I just started by hand on ec2:

* `os.Hostname` returns "ip-172-31-49-127"
* `hostname` returns "ip-172-31-49-127"
* `hostname -f` returns "ip-172-31-49-127.ec2.internal"
* `host 172.31.49.127` (using the IP from `ifconfig`) returns "ip-172-31-49-127.ec2.internal"

So whatever `hostname -f` is doing looks like what we want. Anyone know how that works and how we might get that information from go?

**Comment by knz** [comment]
While looking at #4036

**Comment by knz** [comment]
Can be solved using https://gowalker.org/github.com/tonnerre/golang-dns
Need to provide `ServerName` in `tls.Context`.

**Comment by knz** [comment]
I'm too far from this issue right now to deal with it effectively for Q2/Beta. De-assigning for now. I could put it back on my plate later this year if needed.

**Comment by mberhault** [comment]
Closing. Which address is used depends entirely on the address passed when creating the client.
